File: utils/path_utils.py
from pathlib import Path
from typing import Optional


def validate_directory(path_str: str, *, allowed_base: Optional[Path] = None) -> Path:
    """Return ``Path`` for ``path_str`` if it is a directory within ``allowed_base``.

    Parameters
    ----------
    path_str: str
        Path to validate.
    allowed_base: Optional[Path]
        Optional base directory. If provided, ``path_str`` must resolve inside
        this directory.
    """
    p = Path(path_str).resolve()
    if not p.is_dir():
        raise ValueError(f"{path_str} is not a directory")
    if allowed_base is not None:
        base = allowed_base.resolve()
        try:
            p.relative_to(base)
        except ValueError:
            raise ValueError(f"{p} is outside allowed directory {base}") from None
    return p


def validate_file(path_str: str, *, allowed_base: Optional[Path] = None) -> Path:
    """Return ``Path`` for ``path_str`` if it is a file within ``allowed_base``."""
    p = Path(path_str).resolve()
    if not p.is_file():
        raise ValueError(f"{path_str} is not a file")
    if allowed_base is not None:
        base = allowed_base.resolve()
        try:
            p.relative_to(base)
        except ValueError:
            raise ValueError(f"{p} is outside allowed directory {base}") from None
    return p

File: lib/utils/path-utils.ts
// Path validation utilities
// Port from utils/path_utils.py

function normalizePathString(pathStr: string): string {
  return pathStr.replace(/\\/g, '/').replace(/\/+/g, '/');
}

function hasTraversalSegment(pathStr: string): boolean {
  return normalizePathString(pathStr).split('/').some((segment) => segment === '..');
}

function isAbsolutePath(pathStr: string): boolean {
  const normalized = normalizePathString(pathStr);
  return normalized.startsWith('/') || /^[a-zA-Z]:\//.test(normalized);
}

function isInsideAllowedBase(pathStr: string, allowedBase: string): boolean {
  const normalizedPath = normalizePathString(pathStr).replace(/\/$/, '');
  const normalizedBase = normalizePathString(allowedBase).replace(/\/$/, '');
  return normalizedPath === normalizedBase || normalizedPath.startsWith(`${normalizedBase}/`);
}

function validateSafeRelativePath(pathStr: string, allowedBase?: string): string {
  if (hasTraversalSegment(pathStr) || isAbsolutePath(pathStr)) {
    throw new Error(`Path contains unsafe components: ${pathStr}`);
  }

  if (allowedBase && !isInsideAllowedBase(pathStr, allowedBase)) {
    throw new Error(`Path is outside allowed directory: ${pathStr}`);
  }

  return pathStr;
}

/**
 * Validate that a path string is safe and within allowed directory.
 * Server-side file access should prefer lib/utils/server-paths.ts.
 */
export function validateDirectory(
  pathStr: string,
  allowedBase?: string
): string {
  return validateSafeRelativePath(pathStr, allowedBase);
}

/**
 * Validate that a file path is safe
 */
export function validateFile(
  pathStr: string,
  allowedBase?: string
): string {
  return validateSafeRelativePath(pathStr, allowedBase);
}

/**
 * Sanitize a filename for safe use
 */
export function sanitizeFilename(filename: string): string {
  // Remove or replace unsafe characters
  return filename
    .replace(/[^a-zA-Z0-9_.-]/g, '_')
    .replace(/\.{2,}/g, '.')
    .substring(0, 255); // Limit length
}

/**
 * Check if a path is safe (no traversal attempts)
 */
export function isSafePath(pathStr: string): boolean {
  return !hasTraversalSegment(pathStr) && !isAbsolutePath(pathStr);
}
